tests/0105-transactions_mock.c

```diff
@@ -2866,6 +2866,67 @@ static void do_test_disconnected_group_coord(rd_bool_t switch_coord) {
 }
 
 
+/**
+ * @brief Test that transaction state different from the initial one,
+ * because of a local timed out, doesn't cause an illegal state transition.
+ */
+static void do_test_txn_state_different_from_the_initial_one() {
+        rd_kafka_t *rk;
+        rd_kafka_mock_cluster_t *mcluster;
+        int32_t coord_id = 1;
+        rd_kafka_resp_err_t err;
+        const char *topic            = "test";
+        const char *transactional_id = "txnid";
+        int msgcnt                   = 1000;
+        int remains                  = 0;
+        rd_kafka_error_t *error_commit, *error_abort;
+
+        SUB_TEST_QUICK("Test transaction state different from the initial one");
+
+        rk = create_txn_producer(&mcluster, transactional_id, 3, NULL);
+
+        err = rd_kafka_mock_topic_create(mcluster, topic, 1, 3);
+        TEST_ASSERT(!err, "Failed to create topic: %s", rd_kafka_err2str(err));
+
+        rd_kafka_mock_coordinator_set(mcluster, "transaction", transactional_id,
+                                      coord_id);
+        rd_kafka_mock_partition_set_leader(mcluster, topic, 0, coord_id);
+
+        TEST_SAY("Starting transaction\n");
+        TEST_CALL_ERROR__(rd_kafka_init_transactions(rk, 5000));
+        TEST_CALL_ERROR__(rd_kafka_begin_transaction(rk));
+
+        test_produce_msgs2_nowait(rk, topic, 0, RD_KAFKA_PARTITION_UA, 0,
+                                  msgcnt / 2, NULL, 0, &remains);
+
+        rd_kafka_mock_broker_push_request_error_rtts(
+            mcluster, coord_id, RD_KAFKAP_EndTxn, 1, RD_KAFKA_RESP_ERR_NO_ERROR,
+            70);
+
+        error_commit = rd_kafka_commit_transaction(rk, 50);
+        TEST_SAY_ERROR(error_commit, "commit_transaction() returned: ");
+        TEST_ASSERT(rd_kafka_error_txn_requires_abort(error_commit),
+                    "Expected an abortable error");
+        TEST_ASSERT(rd_kafka_error_code(error_commit) ==
```

Comment:
If I understand correctly what will happen in this case is that the EndTxn(commit) will succeed on the broker, but commit_transaction() will fail locally due to timeout - at which point the application does an abort_transaction().

And we're assuming the abort_transaction() will succeed, which gives the application the idea that the transaction was indeed aborted, but in fact it was successfully committed on the broker - this yields an inconsistent non-repairable view of the last transaction between the cluster and application.

I'm not sure what the proper behaviour is here, but I think it should be that the abort_transaction() fails if there's an in-flight commit. Question is, should it be a retriable, permanent, or fatal error?

What does the Java producer do in this case?

Reply:
Yes it should give an error, not sure if fatal or permanent. I'm checking

Reply:
In Java it's throwing TimeoutException [here](https://github.com/apache/kafka/blob/0507597597e1d2d9adaa33b87e5ea509e12fd2f0/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionalRequestResult.java#L58), that is retriable. So the commit should be retried in that case and not transition not an abortable error.

Reply:
I've pushed a commit to make commits retriable too